### PR TITLE
Add ListRepoNames method to Repo Store

### PR DESCRIPTION
**Pull request by asdine**
This PR adds a new Repo store method called `ListRepoNames` that only lists repo names (and ids) rather than returning the entire repo.
The goal of this method is to be used whenever we don't need to fetch complete repository information (especially important for Search code).

Currently, this method scans 6 fields instead of just the repo ID and Name. Subsequent PRs will change that behaviour from within the `getReposBySQL` method, and unexport the `OnlyRepoIDs` option.

Part of https://github.com/sourcegraph/sourcegraph/issues/16219

### CODE REVIEW

File: internal/db/repos.go [hunk 1]
```diff
@@ -559,10 +581,69 @@ func (s *RepoStore) List(ctx context.Context, opt ReposListOptions) (results []*
...
+
+// ListRepoNames returns a list of repositories names and ids.
+// It overrides the OnlyRepoIDs options by setting it to true.
+func (s *RepoStore) ListRepoNames(ctx context.Context, opt ReposListOptions) (results []*types.RepoName, err error) {
+	tr, ctx := trace.New(ctx, "repos.ListRepoNames", "")
+	defer func() {
+		tr.SetError(err)
+		tr.Finish()
+	}()
+	s.ensureStore()
```
Comment (tsenart) [inline internal/db/repos.go:612]: Just noticed we are not using `sync.Once` in ensureStore. It's much faster than a mutex, we should!

**Review by tsenart — APPROVED** [review]

File: internal/db/repos_db_test.go [hunk 2]
```diff
@@ -1089,6 +1107,681 @@ func TestRepos_List_externalServiceID(t *testing.T) {
...
+			t.Fatal(err)
+		}
+		if got := sortedRepoNames(reposFromRepoNames(repos)); !reflect.DeepEqual(got, test.exp) {
+			t.Errorf("for test case %v, got %v (want %v)", test, got, test.exp)
+		}
+	}
+}
+
+// TestRepos_ListRepoNames_query tests the behavior of Repos.List when called with
+// a query.
```
Comment (unknwon) [inline internal/db/repos_db_test.go:1340]: Needs update? 
```suggestion
// TestRepos_ListRepoNames_query tests the behavior of Repos.ListRepoNames when called with
// a query.
```

File: internal/db/repos_db_test.go [hunk 3]
```diff
@@ -1089,6 +1107,681 @@ func TestRepos_List_externalServiceID(t *testing.T) {
...
+// Test batch 1 (correct filtering)
+func TestRepos_ListRepoNames_query1(t *testing.T) {
```
Comment (unknwon) [inline internal/db/repos_db_test.go:1342]: Why not just name it 
```suggestion
func TestRepos_ListRepoNames_correctFiltering(t *testing.T) {
```

File: internal/db/repos_db_test.go [hunk 4]
```diff
@@ -1089,6 +1107,681 @@ func TestRepos_List_externalServiceID(t *testing.T) {
...
+			t.Fatal(err)
+		}
+		if got := repoNames(reposFromRepoNames(repos)); !reflect.DeepEqual(got, test.want) {
+			t.Errorf("%q: got repos %q, want %q", test.query, got, test.want)
+		}
+	}
+}
+
+// Test batch 2 (correct ranking)
+func TestRepos_ListRepoNames_query2(t *testing.T) {
```
Comment (unknwon) [inline internal/db/repos_db_test.go:1380]: Ditto

**Review by unknwon — COMMENTED** [review]
The critical changes LGTM, but I'm seriously wondering are all additional tests really needed?

If the primary purpose of `ListRepoNames` to not return with all fields, why can't we be confident with only one set of tests?

File: internal/db/repos.go (hunk 1 above)
Comment (asdine) [reply internal/db/repos.go:612]: Good idea!

**Comment by asdine** [comment]
> The critical changes LGTM, but I'm seriously wondering are all additional tests really needed?
> If the primary purpose of ListRepoNames to only return two fields, why can't we be confident with only small set of tests?

@unknwon This is a great point but actually most of these tests will in fact be useful only for `ListNames` rather than `List` because the features tested are only used by the search code.
I initially wanted to merge the tests so that they will test both methods, but we will very likely simplify `List` in the future and remove some of its tests.
So until then, I decided to keep this PR simple by just copying List tests and adapting them to ListNames, and postpone the simplification decision to another PR.

**Review by unknwon — APPROVED** [review]
SGTM

**Pull request merged**